osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs

```diff
@@ -0,0 +1,183 @@
+            AdjustPitch.TriggerChange();
+            recentRates.Clear();
+            recentRates.AddRange(Enumerable.Range(0, average_count).Select(_ => InitialRate.Value));
+        }
+
+        public void ApplyToSample(DrawableSample sample)
+        {
+            sample.AddAdjustment(AdjustableProperty.Frequency, SpeedChange);
+        }
+
+        public double ApplyToRate(double time, double rate = 1) => rate * InitialRate.Value;
+
+        private void applyPitchAdjustment(ValueChangedEvent<bool> adjustPitchSetting)
+        {
+            // remove existing old adjustment
+            track?.RemoveAdjustment(adjustmentForPitchSetting(adjustPitchSetting.OldValue), SpeedChange);
+
+            track?.AddAdjustment(adjustmentForPitchSetting(adjustPitchSetting.NewValue), SpeedChange);
+        }
+
+        private AdjustableProperty adjustmentForPitchSetting(bool adjustPitchSettingValue)
+            => adjustPitchSettingValue ? AdjustableProperty.Frequency : AdjustableProperty.Tempo;
+
+        public void ApplyToDrawableHitObject(DrawableHitObject drawable)
+        {
+            drawable.OnNewResult += (o, result) =>
+            {
+                if (dequeuedRates.ContainsKey(result.HitObject)) return;
+
+                if (!result.IsHit) return;
+                if (!result.Type.AffectsAccuracy()) return;
+                if (!previousEndTimes.ContainsKey(result.HitObject)) return;
+
+                double prevEndTime = previousEndTimes[result.HitObject];
+
+                recentRates.Add(Math.Clamp((result.HitObject.GetEndTime() - prevEndTime) / (result.TimeAbsolute - prevEndTime) * SpeedChange.Value, slowest_rate, fastest_rate));
+
+                dequeuedRates.Add(result.HitObject, recentRates[0]);
+                recentRates.RemoveAt(0);
+
+                overlay.TransformBindableTo(SpeedChange, recentRates.Average(), 100);
+            };
+            drawable.OnRevertResult += (o, result) =>
+            {
+                if (!dequeuedRates.ContainsKey(result.HitObject)) return;
+
+                if (!result.IsHit) return;
+                if (!result.Type.AffectsAccuracy()) return;
+                if (!previousEndTimes.ContainsKey(result.HitObject)) return;
+
+                recentRates.Insert(0, dequeuedRates[result.HitObject]);
+                recentRates.RemoveAt(recentRates.Count - 1);
+                dequeuedRates.Remove(result.HitObject);
+
+                overlay.TransformBindableTo(SpeedChange, recentRates.Average(), 100);
+            };
+        }
+
+        public void ApplyToBeatmap(IBeatmap beatmap)
+        {
+            var hitObjects = getAllApplicableHitObjects(beatmap.HitObjects).ToList();
+            var endTimes = hitObjects.Select(x => x.GetEndTime()).OrderBy(x => x).ToList();
+
+            foreach (HitObject hitObject in hitObjects)
+            {
+                double prevEndTime = endTimes.LastOrDefault(ht => !Precision.AlmostBigger(ht, hitObject.GetEndTime()));
```

Comment:
This is unnecessarily quadratic. Could be improved to O(n log n) by binary searching `endTimes` instead. Several other examples of this exist in this repository.

Reply:
Wow, didn't know about the bitwise complement trick in `List<T>.BinarySearch` at all. What a weird hack.